### ISSUE TITLE
Performance comparison of Agrona

### ISSUE THREAD

**Issue by nmittler** [netty/netty#3613]
Just starting a discussion of performance of the [Agrona](https://github.com/real-logic/Agrona/tree/master/src/main/java/uk/co/real_logic/agrona/collections) [Int2ObjectHashMap](https://github.com/real-logic/Agrona/blob/master/src/main/java/uk/co/real_logic/agrona/collections/Int2ObjectHashMap.java) vs our IntObjectHashMap.

At first look, they blow us away:

| Benchmark |  (size) | Mode | Cnt | Score | Error | Units |
----------------- | -------- | ------- | ---- | ------ | ----- | ----- |
| IntObjectHashMapBenchmark.agronaPut | 10 |  thrpt |    5 |  9656346.576 | ±920367.431 | ops/s | 
| IntObjectHashMapBenchmark.agronaPut | 100 |  thrpt |    5 |  904127.630 | ±94237.691 | ops/s |
| IntObjectHashMapBenchmark.agronaPut | 1000 |  thrpt |    5 |  81677.081 | ±6139.173 | ops/s |
| IntObjectHashMapBenchmark.agronaPut | 10000 |  thrpt |    5 |  6903.031 | ±560.995 | ops/s |
| IntObjectHashMapBenchmark.agronaPut | 100000 |  thrpt |    5 |  460.745 | ±33.215 | ops/s |
| IntObjectHashMapBenchmark.put | 10 |  thrpt |  5 |  6007508.959 | ±1318396.637 | ops/s  | 
| IntObjectHashMapBenchmark.put | 100 |  thrpt |  5 |  638997.389 | ±31698.221 | ops/s |
| IntObjectHashMapBenchmark.put | 1000 |  thrpt |  5 |  54752.095 | ±3773.508 | ops/s |
| IntObjectHashMapBenchmark.put | 10000 |  thrpt |  5 |  4394.468 | ±240.219 | ops/s |
| IntObjectHashMapBenchmark.put | 100000 |  thrpt |  5 |  213.510 | ±34.770 |  ops/s |

**Comment by normanmaurer** [comment]
@nmittler wow :) I guess we have to fix this ;)

**Comment by nmittler** [comment]
@normanmaurer agreed! Any thoughts on just taking Agrona as a dependency? It's pretty small.

**Comment by normanmaurer** [comment]
@nmittler we can't take it for "common" as we want to stay zero-dependency for the core. If its for http2 only I have no problem with it to be honest. @trustin wdyt ?

**Comment by ejona86** [comment]
Agrona requires Java 8, FWIW. I don't know why, but both the README.md and gradle.build say 1.8.

**Comment by nmittler** [comment]
@ejona86 yeah I saw that ... worst case we could replace our implementation with theirs.

**Comment by buchgr** [comment]
@nmittler thanks for the benchmark! Very interesting results!

I ll extend the benchmark to
 1. See how we compare to other collection libraries esp. JDK, fastutil and Koloboke. I have no idea if Agrona is particularly fast.
 2. Retrieval perf will be interesting
 3. Different number distributions. Random numbers are fine, but I would argue that they are the best case for hash maps as they naturally avoid collisions and don't test the hash function too much. I think a very interesting evaluation would be to see how performance changes when the numbers follow some distribution / pattern (e.g. 2,4,6,8 like we see with HTTP/2 streams). Tests like this will really tell the quality of the hash function and collision handling.

**Comment by Scottmitch** [comment]
@nmittler - Thanks for sharing!  Room for improvement for sure :)

@buchgr - +1 for your extension cases.  Also what about inserting the same key multiple times.

**Comment by Scottmitch** [comment]
@ejona86 - Thanks for pointing that out.  I think java 8 is a deal breaker.  We need java 7 (and I think java 6) for 4.x.

@nmittler - If there is java8 stuff...maybe we could port over to supplement these with older JDK stuff?  Either way we should wait until @buchgr adds other libraries to see if there is a better one before we adopt one....

**Comment by normanmaurer** [comment]
+1

**Comment by nmittler** [comment]
@Scottmitch +1

**Comment by trustin** [comment]
If their (the fastest one for our use cases) license is compatible we could fork the necessary portion only after doing a backport if necessary.

We'll have to keep track of the forked works from time to time though, so we athe fixed from the upstream.

Alternatively, we could consider changing our policy about the core dependencies in 5. Ideas?

**Comment by nmittler** [comment]
If we did decide to take an extra dependency, it wouldn't need to be on the common module.  I believe that the map is used mostly by the http2 module at present. The HTTP headers also uses it, but depending on the aftermath of #3600, this may change as well.

**Comment by nmittler** [comment]
FYI, I just changed (locally) the map template to use masking rather than mod for generating the hash index and our performance is now on par with Agrona.  Bear in mind these numbers were generated on my laptop :)

Benchmark                         (keyDistribution)  (mapType)  (size)   Mode  Cnt         Score         Error  Units
IntObjectHashMapBenchmark.lookup              HTTP2     AGRONA      10  thrpt   20  13620483.951 ±  229904.656  ops/s
IntObjectHashMapBenchmark.lookup              HTTP2     AGRONA     100  thrpt   20   1305585.600 ±   34508.157  ops/s
IntObjectHashMapBenchmark.lookup              HTTP2     AGRONA    1000  thrpt   20    139656.520 ±     658.561  ops/s
IntObjectHashMapBenchmark.lookup              HTTP2     AGRONA   10000  thrpt   20     13544.406 ±      71.935  ops/s
IntObjectHashMapBenchmark.lookup              HTTP2     AGRONA  100000  thrpt   20      1119.435 ±      11.037  ops/s
IntObjectHashMapBenchmark.lookup              HTTP2      NETTY      10  thrpt   20  13957628.409 ±   81318.355  ops/s
IntObjectHashMapBenchmark.lookup              HTTP2      NETTY     100  thrpt   20   1423707.183 ±   17882.081  ops/s
IntObjectHashMapBenchmark.lookup              HTTP2      NETTY    1000  thrpt   20    141023.125 ±    1442.523  ops/s
IntObjectHashMapBenchmark.lookup              HTTP2      NETTY   10000  thrpt   20     13871.854 ±     176.663  ops/s
IntObjectHashMapBenchmark.lookup              HTTP2      NETTY  100000  thrpt   20      1447.520 ±      15.963  ops/s

IntObjectHashMapBenchmark.lookup             RANDOM     AGRONA      10  thrpt   20  12948472.879 ± 1441240.814  ops/s
IntObjectHashMapBenchmark.lookup             RANDOM     AGRONA     100  thrpt   20   1385976.997 ±   19167.139  ops/s
IntObjectHashMapBenchmark.lookup             RANDOM     AGRONA    1000  thrpt   20    123026.186 ±    2099.188  ops/s
IntObjectHashMapBenchmark.lookup             RANDOM     AGRONA   10000  thrpt   20      8820.420 ±     263.282  ops/s
IntObjectHashMapBenchmark.lookup             RANDOM     AGRONA  100000  thrpt   20       652.950 ±      11.470  ops/s
IntObjectHashMapBenchmark.lookup             RANDOM      NETTY      10  thrpt   20  12745496.349 ±  477420.478  ops/s
IntObjectHashMapBenchmark.lookup             RANDOM      NETTY     100  thrpt   20   1287580.442 ±   10140.038  ops/s
IntObjectHashMapBenchmark.lookup             RANDOM      NETTY    1000  thrpt   20    121394.509 ±    6362.047  ops/s
IntObjectHashMapBenchmark.lookup             RANDOM      NETTY   10000  thrpt   20      9546.140 ±     253.843  ops/s
IntObjectHashMapBenchmark.lookup             RANDOM      NETTY  100000  thrpt   20       777.330 ±       9.389  ops/s

IntObjectHashMapBenchmark.put                 HTTP2     AGRONA      10  thrpt   20   4038745.496 ±   23410.951  ops/s
IntObjectHashMapBenchmark.put                 HTTP2     AGRONA     100  thrpt   20    471491.156 ±    8327.760  ops/s
IntObjectHashMapBenchmark.put                 HTTP2     AGRONA    1000  thrpt   20     54346.160 ±     315.747  ops/s
IntObjectHashMapBenchmark.put                 HTTP2     AGRONA   10000  thrpt   20      4446.494 ±      50.799  ops/s
IntObjectHashMapBenchmark.put                 HTTP2     AGRONA  100000  thrpt   20       371.513 ±       5.281  ops/s
IntObjectHashMapBenchmark.put                 HTTP2      NETTY      10  thrpt   20   6304389.260 ±   75939.511  ops/s
IntObjectHashMapBenchmark.put                 HTTP2      NETTY     100  thrpt   20    633409.353 ±    4056.028  ops/s
IntObjectHashMapBenchmark.put                 HTTP2      NETTY    1000  thrpt   20     78958.278 ±     672.576  ops/s
IntObjectHashMapBenchmark.put                 HTTP2      NETTY   10000  thrpt   20      6005.443 ±      72.580  ops/s
IntObjectHashMapBenchmark.put                 HTTP2      NETTY  100000  thrpt   20       657.625 ±       3.543  ops/s

IntObjectHashMapBenchmark.put                RANDOM     AGRONA      10  thrpt   20   4253496.089 ±   43866.873  ops/s
IntObjectHashMapBenchmark.put                RANDOM     AGRONA     100  thrpt   20    507736.519 ±    7035.085  ops/s
IntObjectHashMapBenchmark.put                RANDOM     AGRONA    1000  thrpt   20     53597.090 ±     537.097  ops/s
IntObjectHashMapBenchmark.put                RANDOM     AGRONA   10000  thrpt   20      2290.315 ±      18.757  ops/s
IntObjectHashMapBenchmark.put                RANDOM     AGRONA  100000  thrpt   20       221.391 ±       2.870  ops/s
IntObjectHashMapBenchmark.put                RANDOM      NETTY      10  thrpt   20   5541931.509 ±   53905.160  ops/s
IntObjectHashMapBenchmark.put                RANDOM      NETTY     100  thrpt   20    549707.881 ±    6451.897  ops/s
IntObjectHashMapBenchmark.put                RANDOM      NETTY    1000  thrpt   20     66371.368 ±     485.428  ops/s
IntObjectHashMapBenchmark.put                RANDOM      NETTY   10000  thrpt   20      2545.250 ±      16.603  ops/s
IntObjectHashMapBenchmark.put                RANDOM      NETTY  100000  thrpt   20       257.833 ±       1.838  ops/s

IntObjectHashMapBenchmark.remove              HTTP2     AGRONA      10  thrpt   20   2548389.149 ±   16950.027  ops/s
IntObjectHashMapBenchmark.remove              HTTP2     AGRONA     100  thrpt   20    324296.724 ±    4544.768  ops/s
IntObjectHashMapBenchmark.remove              HTTP2     AGRONA    1000  thrpt   20     31260.378 ±     336.061  ops/s
IntObjectHashMapBenchmark.remove              HTTP2     AGRONA   10000  thrpt   20      2695.204 ±      32.257  ops/s
IntObjectHashMapBenchmark.remove              HTTP2     AGRONA  100000  thrpt   20       266.294 ±       3.432  ops/s
IntObjectHashMapBenchmark.remove              HTTP2      NETTY      10  thrpt   20   3630902.585 ±  198373.508  ops/s
IntObjectHashMapBenchmark.remove              HTTP2      NETTY     100  thrpt   20    379713.731 ±    3743.644  ops/s
IntObjectHashMapBenchmark.remove              HTTP2      NETTY    1000  thrpt   20     44098.505 ±     425.800  ops/s
IntObjectHashMapBenchmark.remove              HTTP2      NETTY   10000  thrpt   20      4112.791 ±      29.246  ops/s
IntObjectHashMapBenchmark.remove              HTTP2      NETTY  100000  thrpt   20       441.965 ±       6.063  ops/s

IntObjectHashMapBenchmark.remove             RANDOM     AGRONA      10  thrpt   20   2417231.641 ±   50854.739  ops/s
IntObjectHashMapBenchmark.remove             RANDOM     AGRONA     100  thrpt   20    273529.259 ±    5248.752  ops/s
IntObjectHashMapBenchmark.remove             RANDOM     AGRONA    1000  thrpt   20     24310.217 ±     841.650  ops/s
IntObjectHashMapBenchmark.remove             RANDOM     AGRONA   10000  thrpt   20      1333.508 ±      10.388  ops/s
IntObjectHashMapBenchmark.remove             RANDOM     AGRONA  100000  thrpt   20       135.364 ±       0.874  ops/s
IntObjectHashMapBenchmark.remove             RANDOM      NETTY      10  thrpt   20   3258849.693 ±   17954.305  ops/s
IntObjectHashMapBenchmark.remove             RANDOM      NETTY     100  thrpt   20    329100.862 ±    3481.484  ops/s
IntObjectHashMapBenchmark.remove             RANDOM      NETTY    1000  thrpt   20     33187.179 ±    2590.791  ops/s
IntObjectHashMapBenchmark.remove             RANDOM      NETTY   10000  thrpt   20      1571.308 ±      11.714  ops/s
IntObjectHashMapBenchmark.remove             RANDOM      NETTY  100000  thrpt   20       152.300 ±       3.653  ops/s

**Comment by buchgr** [comment]
@nmittler thanks looks good. With a size of 100k elements we will have roughly 400k of data. So we have a ~30% chance of hitting L1, and depending on your processor it might all fit in L2. That way collisions don't really hit us that hard. It would be interesting to see how everything performs with 1M or 10M elements.

**Comment by Scottmitch** [comment]
@nmittler - masking for the win!  Looks great... I'll take a look at https://github.com/netty/netty/pull/3627

BTW I wrapped your post in 3 ticks so it is a bit easier to read.

**Comment by nmittler** [comment]
@Scottmitch  thanks, much better! :)

**Comment by nmittler** [comment]
@buchgr @Scottmitch now that #3627 is in, I think we can close this.  WDYT?

**Comment by Scottmitch** [comment]
I'm satisfied.  Not sure if @buchgr  was interested in a follow-up issue comparing other "high performance" native type collections?

**Comment by buchgr** [comment]
@Scottmitch yeah I already did a few days ago. Haven't had time to share them publicly yet, but discussed results with @nmittler . Hope I ll be able to share them over the next few days, but looks good. So it's fine to close this.

**Comment by nmittler** [comment]
ok great... closing. Thanks guys!

**Comment by buchgr** [comment]
@nmittler @Scottmitch here are the results I got a few days back. I made some changes to the benchmarking code and I didn't have time to clean up the code which is why I didn't wanna share the results yet (without the code) https://gist.github.com/buchgr/6461a7e4836c37894014

**Comment by nmittler** [comment]
@buchgr thanks!  Looks like FastUtil is the winner overall, but we're definitely in much better shape.

**Comment by Scottmitch** [comment]
@buchgr Thanks!  I left a comment on the gist https://gist.github.com/buchgr/6461a7e4836c37894014#comment-1434988.
